### PR TITLE
cql3: Fix reverse-order collections

### LINKED ISSUE #7868: Assertion failure (crash) when clustering key is a frozen collection and reverse order
The following trivial cql-pytest, creating a table with a frozen collection and reverse sort order and adding one item to it with a **prepared statement**, crashes Scylla on an assertion failure during the insert:
```python
def test_xyz(cql, test_keyspace):
    with create_table(cql, test_keyspace, f"(a int, b frozen<set<int>>, PRIMARY KEY (a,b)) WITH CLUSTERING ORDER BY (b DESC)") as table:
        execute(cql, table, "INSERT INTO %s (a, b) VALUES (?, ?)", 0, {1, 2, 3});
```

(Note: in this code, the `execute` convenience function does both prepare and execute steps - and it is actually the prepare step which fails, not the execute).

The assertion failure is:
```
test_scylla: cql3/abstract_marker.cc:80: virtual ::shared_ptr<term> cql3::abstra
ct_marker::raw::prepare(database &, const seastar::sstring &, lw_shared_ptr<cql3
::column_specification>) const: Assertion `0' failed.
```

The `WITH CLUSTERING ORDER BY (b DESC)` is important for the crash, it doesn't happen without it. Also, it seems to reproduce with frozen collections of any type but not with, for example, int. The content of the set (here `{1, 2, 3}`) doesn't matter - an empty set `{}` will crash the same.

This crash is also specific to prepared statements - it happens in the prepare phase. Unprepared statements also have a bug in this case, but it looks different: I opened a separate issue about it - #7875.

The backtrace of the above assertion failure is:
```
cql3::abstract_marker::raw::prepare(database&, seastar::basic_sstring<char, unsigned int, 15u, true> const&, seastar::lw_shared_ptr<cql3::column_specification>) const at abstract_marker.cc:?
cql3::single_column_relation::to_term(std::vector<seastar::lw_shared_ptr<cql3::column_specification>, std::allocator<seastar::lw_shared_ptr<cql3::column_specification> > > const&, cql3::term::raw const&, database&, seastar::basic_sstring<char, unsigned int, 15u, true> const&, cql3::variable_specifications&) const at single_column_relation.cc:?
cql3::single_column_relation::new_EQ_restriction(database&, seastar::lw_shared_ptr<schema const>, cql3::variable_specifications&) at single_column_relation.cc:?
cql3::relation::to_restriction(database&, seastar::lw_shared_ptr<schema const>, cql3::variable_specifications&) at relation.cc:?
cql3::restrictions::statement_restrictions::statement_restrictions(database&, seastar::lw_shared_ptr<schema const>, cql3::statements::statement_type, std::vector<seastar::shared_ptr<cql3::relation>, std::allocator<seastar::shared_ptr<cql3::relation> > > const&, cql3::variable_specifications&, bool, bool, bool, bool) at statement_restrictions.cc:?
cql3::statements::modification_statement::process_where_clause(database&, std::vector<seastar::shared_ptr<cql3::relation>, std::allocator<seastar::shared_ptr<cql3::relation> > >, cql3::variable_specifications&) at modification_statement.cc:?
cql3::statements::raw::insert_statement::prepare_internal(database&, seastar::lw_shared_ptr<schema const>, cql3::variable_specifications&, std::unique_ptr<cql3::attributes, std::default_delete<cql3::attributes> >, cql3::cql_stats&) const at update_statement.cc:?
cql3::statements::raw::modification_statement::prepare(database&, cql3::variable_specifications&, cql3::cql_stats&) const at modification_statement.cc:?
cql3::statements::raw::modification_statement::prepare(database&, cql3::cql_stats&) at modification_statement.cc:?
```

CC @dekimir who might be familiar with the relations which the backtrace suggests might be involved (but I don't know yet if it is).

This bug also crashes the Cassandra unit test which I translated into Python, `cassandra_tests/validation/entities/frozen_collections_test.py`::`testClusteringKeyUsageWithReverseOrderSet`.

### LINKED ISSUE #7875: Failed unprepared insert when clustering key is a frozen collection and reverse order
In issue #7868 I noted that in a table with a frozen collection and reverse sort order, when you add one item to it with a prepared statement, Scylla crashes on an assertion failure. 

We also have a different bug if we try to add an item to such a table with an **unprepared statement**, and this issue is about this bug.

The following test code:
```python
def test_clustering_key_reverse_frozen_collection(cql, test_keyspace):
    with new_test_table(cql, test_keyspace,
            "a int, b frozen<set<int>>, PRIMARY KEY (a,b)",
            "WITH CLUSTERING ORDER BY (b DESC)") as table:
         cql.execute("INSERT INTO " + table + " (a, b) VALUES (0, {1, 2, 3})")

```

Results in an unjustified error message "Invalid set literal for b of type frozen<set<int>>".
There is no such error if we drop the reverse ordering ("WITH CLUSTERING ORDER"), and Cassandra also accepts this case.

### PR DESCRIPTION
**Pull request by dekimir**
When a clustering order is reversed, the corresponding column type is set to reversed_type_impl.  This was not properly handled by collections code, which assumed the type is a regular collection implementation.  Fixed by getting the underlying type when appropriate.

Fixes #7868
Fixes #7875

Tests: unit (dev)

### CODE REVIEW

**Comment by dekimir** [comment]
@nyh I added my own test, since I couldn't apply your patch to master.  Sorry. :(

**Comment by psarna** [comment]
(I took liberty of adding `Fixes #7868` to the description, since that would get translated to the merged commit message - it's easier to backport if the merge message also mentions the issue.)

**Review by psarna — APPROVED** [review]
Looks good to me, but from the discussion I conclude that there's another conflicting patch which adds tests, so I'll wait for @nyh to confirm

File: cql3/abstract_marker.cc [hunk 1]
```diff
@@ -69,18 +69,19 @@ abstract_marker::raw::raw(int32_t bind_index)
 
 ::shared_ptr<term> abstract_marker::raw::prepare(database& db, const sstring& keyspace, lw_shared_ptr<column_specification> receiver) const
 {
-    if (receiver->type->is_collection()) {
-        if (receiver->type->get_kind() == abstract_type::kind::list) {
+    const auto tp = receiver->type->underlying_type();
```
Comment (nyh) [inline cql3/abstract_marker.cc:72]: Wow. So potentially we have such type->get_kind() calls returning the wrong thing for reversed collections all over the code. Would be good if you or someone else could audit for that...
I'm curious - why can't type->get_kind() return the "kind" of the underlying type, to solve this problem for all callers?

Comment (nyh) [reply cql3/abstract_marker.cc:72]: Also, a question: I understand what underlying_type does for reversed-order types. Is it possible that for other things, it has a different meaning, so it wouldn't be correct to use underlying_type?

**Comment by nyh** [comment]
Oh, and please check using my v2 tests if your patch also fixes #7875, and if not, consider a patch which fixes both bugs (if it's not the same problem, it's probably almost the same fix, with the underlying_type thing).

File: test/cql-pytest/cassandra_tests/validation/entities/collections_test.py [hunk 2]
```diff
@@ -1583,3 +1583,10 @@ def testSelectionOfEmptyCollections(cql, test_keyspace):
             assert_rows(execute(cql, table, "SELECT m['0'], s[0] FROM %s WHERE k = 3"), [None, None])
             assert_rows(execute(cql, table, "SELECT m['0'..'1'], s[0..1] FROM %s WHERE k = 3"), [None, None])
             assert_rows(execute(cql, table, "SELECT m['0'..'1']['3'..'5'], s[0..1][3..5] FROM %s WHERE k = 3"), [None, None])
+
+# Regression-test for #7868.
+def test_7868(cql, test_keyspace):
```
[the remark this answers is not in the document]
Comment (dekimir) [reply test/cql-pytest/cassandra_tests/validation/entities/collections_test.py:1588]: Use it how?  Wait for it to land in master?  Or wholy incorporate it here?

File: cql3/abstract_marker.cc (hunk 1 above)
Comment (dekimir) [reply cql3/abstract_marker.cc:72]: > Is it possible that for other things, it has a different meaning, so it wouldn't be correct to use underlying_type?

That's definitely possible, which is why I had to add the extra is_reversed() check in lists.cc.  I'll add it everywhere, just to be sure.

File: test/cql-pytest/cassandra_tests/validation/entities/collections_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nyh) [reply test/cql-pytest/cassandra_tests/validation/entities/collections_test.py:1588]: I meant apply it locally just to see my tests pass with your new code. You don't need to include it inside your patch.
But whatever you do, you can't add new tests to the cassandra_tests/ subdirectory - those are supposed to be translated tests, not new tests :-)

**Comment by dekimir** [comment]
> Oh, and please check using my v2 tests if your patch also fixes #7875

Already checked; it doesn't.

> and if not, consider a patch which fixes both bugs

I considered it and rejected it, because I want to prioritize crashes over lesser bugs.  Crashes should be fixed promptly, without waiting for extra work elsewhere.

**Comment by nyh** [comment]
> > Oh, and please check using my v2 tests if your patch also fixes #7875
> 
> Already checked; it doesn't.
> 
> > and if not, consider a patch which fixes both bugs
> 
> I considered it and rejected it, because I want to prioritize crashes over lesser bugs. Crashes should be fixed promptly, without waiting for extra work elsewhere.

I think this is a mistake... You're already thinking about this code, I already wrote tests for it - why not just fix it? The fix is probably (but I'm just guessing) yet another test for "kind" which should look at underlying_type...

**Comment by dekimir** [comment]
> > I considered it and rejected it, because I want to prioritize crashes over lesser bugs. Crashes should be fixed promptly, without waiting for extra work elsewhere.
> 
> I think this is a mistake... You're already thinking about this code, I already wrote tests for it - why not just fix it?

Because it's not this code, it's some other code that I want to debug separately from this.  I have a fix for the crash, and I don't understand why that should wait for some other work in another part of the code.

File: test/cql-pytest/cassandra_tests/validation/entities/collections_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dekimir) [reply test/cql-pytest/cassandra_tests/validation/entities/collections_test.py:1588]: Good point about translated-vs-new tests.  But I do want to have _a_ test with the code change.  Either your test patch has to get in (so mine will just remove `@skip`) or the two should be united into one.

**Comment by nyh** [comment]
> > I think this is a mistake... You're already thinking about this code, I already wrote tests for it - why not just fix it?
> 
> Because it's not this code, it's some other code that I want to debug separately from this. I have a fix for the crash, and I don't understand why that should wait for some other work in another part of the code.

During the hours we have spent on writing these tests and debugging the problem, we discovered a serial bug - that every piece of code that looks if "type->kind == list" will miss reversed collections. There may be a dozen such bugs lurking in the program, but we know of at least *two*. If you don't want to spend your time look for all of them, you should at least look at the two we know about. The second one is `sets::literal::validate_assignable_to` and friends, which also check type without considering a reversed type's underlying_type.

There are two reasons why I insist that you look at both problems together:

1. Since we already did 99% of the work of finding the cause and writing tests, let's finish it!
2. After looking at the two problems, maybe this will influence how you think we should solve this problem. Maybe we should have some function that both pieces of code will use?

I think that making "reversed" a wrapper type (done 5 years ago in commit 1047f36a9a391ad75bd127903928309bbf21c0f7) might have been a mistake, it complicates all other code which needs to understand the real type. I'm not asking you to undo this, but I'm asking you to think a little bit further beyond fixing a specific crash.


It's not "some other work in another part of the code". I am guessing (and if you don't even look, you can't disprove my guess) th

**Comment by dekimir** [comment]
@nyh I honestly didn't think it would take so many hours to commit a clean crash fix.  But now that it's happened, I guess I'll spend some more time to fix both bugs in one patchset and delay my other work. :man_shrugging:

**Comment by nyh** [comment]
> @nyh I honestly didn't think it would take so many hours to commit a clean crash fix. But now that it's happened, I guess I'll spend some more time to fix both bugs in one patchset and delay my other work. man_shrugging

You feel this way just because you are underestimating the amount of work that *I* did to discover, document, and test this bug. These are not two "unrelated" bugs - they are places in the code with exactly the same bug - checking "type" without considering underlying() as well. Since we already spent so much time to understand this, now is the best time to fix both places, not just one... (and seeing your objection, I'll not insist that you look for more instances of the same bug, though I suspect there are...).

File: test/cql-pytest/cassandra_tests/validation/entities/collections_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dekimir) [reply test/cql-pytest/cassandra_tests/validation/entities/collections_test.py]: In v2, I didn't add any tests (though I tested with the tests from the mailing list).

File: cql3/abstract_marker.cc (hunk 1 above)
Comment (dekimir) [reply cql3/abstract_marker.cc]: I added two utility methods to abstract_type: self_or_reversed() and as().  But they must be called explicitly.

**Comment by dekimir** [comment]
I pushed v2, PTAL.  I used the tests from the mailing list, though I didn't include them in this PR.  (All the tests pass, btw.)

As agreed, I didn't hunt down all potential bugs from unhandled reversed_type_impl, just the ones necessary to pass the tests.

In v2:
- fixed #7875
- added utility methods to abstract_type that handle reversed types under the hood
- dropped my tests (will rely on Nadav's instead)

**Review by nyh — APPROVED** [review]
Thanks, looks good. I'll test and if all goes well, commit.
We should remember that after this commit, still two potential problems remain:
1. Although you got rid of a couple of asserts() and replaced them by an exception (thanks!), I see there are still a few asserts() missing. I don't know how to trigger them now, but there's is potential that maliciously crafted junk in the CQL protocol can lead to them. Eventually we should get rid of these asserts, but it's not urgent now.
2. As you also also state, there may be additional existing code which checks is_set() and this patch doesn't fix and we didn't yet find in tests. Eventually we may want to look into that, but doesn't need to happen now.

**Comment by psarna** [comment]
What's the status of this pull request? It hasn't been updated in a while and I see some merge conflicts.

**Comment by dekimir** [comment]
@psarna It was merged and should have been closed automatically.

**Comment by avikivity** [comment]
Merged in 176ff0238abba11cf0935cf09f71a363b25aab58. A bad rebase killed the merge commit, but the meat is there.

**Pull request closed without merging**